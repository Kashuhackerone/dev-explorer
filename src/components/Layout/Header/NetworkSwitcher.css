#header-network-dropdown {
  margin: 0 0.5rem;
  background-color: white;
  border: solid 1px rgba(0, 0, 0, 0.4);
  border-radius: 4px;
}

.info-icon {
  align-self: center;
  color: darkgrey;
}

.custom-navbar .dropdown-menu {
  min-width: 200px;
}

.default-node-div {
  height: 32px;
  display: flex;
  justify-content: space-between;
}

.default-node-item {
  padding: 0.25rem 0rem 0.25rem 1.5rem;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 175px;
}

.minus-icon-item {
  align-self: center;
  width: 2rem;
  margin: 0 0.5rem;
  padding: 0.5rem
}

.add-node-div {
  display: flex;
  margin-top: 1rem;
  padding-left: 1rem;
}

.plus-icon-item {
  width: 2rem;
  margin: 0 0.5rem;
  padding: 0.5rem;
}