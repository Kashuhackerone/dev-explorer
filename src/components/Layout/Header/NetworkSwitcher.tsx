import React, { useState, useContext, useEffect, useCallback } from 'react'
import { Nav, NavDropdown, Tooltip, OverlayTrigger, Form } from 'react-bootstrap'
import { useHistory } from 'react-router-dom'

import { defaultNetworks, useNetworkName, useNetworkUrl } from 'src/services/network/networkProvider'
import { UserPrefContext } from 'src/services/userPref/userPrefProvider'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faInfoCircle, faPlus, faMinus } from '@fortawesome/free-solid-svg-icons'

import './NetworkSwitcher.css'

const NetworkSwitcher: React.FC = () => {

  const history = useHistory()
  const networkName = useNetworkName()
  const networkUrl = useNetworkUrl()

  const userPrefContext = useContext(UserPrefContext)
  const { nodeUrlMap, setNodeUrlMap } = userPrefContext!

  const [newNode, setNewNode] = useState('')
  const [showDropdown, setShowDropdown] = useState(false)
  const [currentNetwork, setCurrentNetwork] = useState(defaultNetworks[networkUrl] || networkUrl)

  useEffect(() => {
    setCurrentNetwork(networkName)
  }, [networkName])

  const changeNetwork = useCallback((k: string) => {
    if (k === Object.keys(defaultNetworks)[0])
      history.push('/')
    else
      history.push({
        pathname: '/',
        search: '?' + new URLSearchParams({ network: k }).toString()
      })
  }, [history])

  const addNewNode = () => {
    if (!Object.keys({ ...defaultNetworks, ...nodeUrlMap }).includes(newNode)) {
      nodeUrlMap[newNode] = newNode
      setNodeUrlMap(nodeUrlMap)
    }
    changeNetwork(newNode)
    setNewNode('')
    setShowDropdown(false)
  }

  const deleteNode = (k: string) => {
    delete nodeUrlMap[k]
    setNodeUrlMap(nodeUrlMap)
    changeNetwork(Object.keys(defaultNetworks)[0])
    setShowDropdown(false)
  }

  return (
    <Nav>
      <OverlayTrigger placement='left'
        overlay={<Tooltip id={'network-tt'}> {networkUrl} </Tooltip>}>
        <FontAwesomeIcon className='info-icon' icon={faInfoCircle} />
      </OverlayTrigger>
      <NavDropdown onToggle={(e: boolean) => { setShowDropdown(e) }}
        show={showDropdown} title={currentNetwork} id="header-network-dropdown">
        {Object.entries(defaultNetworks).map(([k, v]) => (
          <NavDropdown.Item key={k} onClick={() => {
            if (currentNetwork !== v) {
              changeNetwork(k)
            }
          }}>
            {v}
          </NavDropdown.Item>
        ))}
        <NavDropdown.Divider />
        {Object.entries(nodeUrlMap).map(([k, v]) => (
          <div key={k} className='default-node-div'>
            <NavDropdown.Item className='default-node-item' onClick={() => {
              if (currentNetwork !== v) {
                changeNetwork(k)
              }
            }}>
              {v}
            </NavDropdown.Item>
            <NavDropdown.Item className='minus-icon-item' onClick={() => { deleteNode(k) }}>
              <FontAwesomeIcon size='lg' icon={faMinus} />
            </NavDropdown.Item>
          </div>
        ))}
        <div className='add-node-div'>
          <Form onSubmit={(e: React.SyntheticEvent) => { e.preventDefault(); addNewNode() }}>
            <Form.Control
              type="text"
              value={newNode}
              onChange={(e) => { setNewNode(e.target.value) }}
              placeholder="Add Network" />
          </Form>
          <NavDropdown.Item className='plus-icon-item' onClick={addNewNode}>
            <FontAwesomeIcon size='lg' icon={faPlus} />
          </NavDropdown.Item>
        </div>
      </NavDropdown>
    </Nav>
  )
}

export default NetworkSwitcher
