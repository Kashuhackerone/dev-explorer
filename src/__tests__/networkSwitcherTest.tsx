import * as React from 'react'
import { shallow } from 'enzyme'

import NetworkSwitcher from 'src/components/Layout/Header/NetworkSwitcher'
import { Form } from 'react-bootstrap'

const mockHistoryPush = jest.fn()

jest.mock("react-router-dom", () => {
  const orig = jest.requireActual("react-router-dom")
  return {
    ...orig,
    useHistory: () => ({
      push: mockHistoryPush
    })
  }
})

jest.mock("src/services/network/networkProvider", () => {
  const orig = jest.requireActual("src/services/network/networkProvider")
  return {
    ...orig,
    defaultNetworks: {
      'https://api.zilliqa.com': 'Mainnet',
      'https://dev-api.zilliqa.com': 'Testnet',
      'https://zilliqa-isolated-server.zilliqa.com': 'Isolated Server',
      'http://52.187.126.172:4201': 'Mainnet Staked Seed Node'
    },
    useNetworkName: () => ('Mainnet'),
    useNetworkUrl: () => ('https://api.zilliqa.com'),
  }
})

describe('<NetworkSwitcher />', () => {
  const mockSetNodeUrlMap = jest.fn()

  beforeEach(() => {
    jest.spyOn(React, 'useContext').mockImplementation(() => ({
      nodeUrlMap: {},
      setNodeUrlMap: mockSetNodeUrlMap
    }))
  })

  it('renders with a given list of default networks and switches correctly', () => {
    const wrapper = shallow(<NetworkSwitcher />)
    expect(wrapper).toMatchSnapshot()

    wrapper.findWhere(node => node.key() === 'https://dev-api.zilliqa.com').simulate('click')

    expect(mockHistoryPush).toHaveBeenCalledWith({ "pathname": "/", "search": "?network=https%3A%2F%2Fdev-api.zilliqa.com" })
  })

  it('adds new network url correctly', () => {
    const wrapper = shallow(<NetworkSwitcher />)
    const inputUrlForm = wrapper.find(Form)
    inputUrlForm.simulate('submit', {
      preventDefault: jest.fn(),
      target: { name: "networkUrl", value: "spam" }
    })
    expect(mockSetNodeUrlMap).toHaveBeenCalledTimes(1)
  })
})