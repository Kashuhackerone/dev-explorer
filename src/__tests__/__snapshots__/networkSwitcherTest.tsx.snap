// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NetworkSwitcher /> renders with a given list of default networks and switches correctly 1`] = `
<Nav
  fill={false}
  justify={false}
  style={
    Object {
      "minWidth": "120px",
    }
  }
>
  <OverlayTrigger
    defaultShow={false}
    overlay={
      <ForwardRef
        id="network-tt"
        placement="right"
      >
         
        https://api.zilliqa.com
         
      </ForwardRef>
    }
    placement="left"
    trigger={
      Array [
        "hover",
        "focus",
      ]
    }
  >
    <FontAwesomeIcon
      border={false}
      className="info-icon"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f05a",
            "M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z",
          ],
          "iconName": "info-circle",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </OverlayTrigger>
  <NavDropdown
    id="header-network-dropdown"
    onToggle={[Function]}
    show={false}
    title="Mainnet"
  >
    <DropdownItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "SafeAnchor",
          "render": [Function],
        }
      }
      disabled={false}
      key="https://api.zilliqa.com"
      onClick={[Function]}
    >
      Mainnet
    </DropdownItem>
    <DropdownItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "SafeAnchor",
          "render": [Function],
        }
      }
      disabled={false}
      key="https://dev-api.zilliqa.com"
      onClick={[Function]}
    >
      Testnet
    </DropdownItem>
    <DropdownItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "SafeAnchor",
          "render": [Function],
        }
      }
      disabled={false}
      key="https://zilliqa-isolated-server.zilliqa.com"
      onClick={[Function]}
    >
      Isolated Server
    </DropdownItem>
    <DropdownItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "SafeAnchor",
          "render": [Function],
        }
      }
      disabled={false}
      key="http://52.187.126.172:4201"
      onClick={[Function]}
    >
      Mainnet Staked Seed Node
    </DropdownItem>
    <DropdownDivider
      role="separator"
    />
    <div
      className="add-node-div"
    >
      <Form
        inline={false}
        onSubmit={[Function]}
      >
        <FormControl
          onChange={[Function]}
          placeholder="Add Network"
          type="text"
          value=""
        />
      </Form>
      <DropdownItem
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "SafeAnchor",
            "render": [Function],
          }
        }
        className="plus-icon-item"
        disabled={false}
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f067",
                "M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z",
              ],
              "iconName": "plus",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size="lg"
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </DropdownItem>
    </div>
  </NavDropdown>
</Nav>
`;
